src/ui/src/core/navigation.ts

```diff
@@ -0,0 +1,70 @@
+type ParsedHash = {
+	pageKey?: string;
+	routeVars: Map<string, string>; // Stored as Map to avoid injection e.g. prototype pollution
+};
+
+const hashRegex = /^((?<pageKey>[^/]*))?(\/(?<routeVars>.*))?$/;
+const routeVarRegex = /^(?<key>[^=]+)=(?<value>.*)$/;
+
+export function getParsedHash(): ParsedHash {
+	const docHash = document.location.hash.substring(1);
+	const hashMatchGroups = docHash.match(hashRegex)?.groups;
+	const routeVars: Map<string, string> = new Map();
+	const pageKey = hashMatchGroups?.pageKey
+		? decodeURIComponent(hashMatchGroups.pageKey)
+		: undefined;
+
+	if (!hashMatchGroups) return { pageKey, routeVars };
+
+	const routeVarsSegments = hashMatchGroups.routeVars?.split("&") ?? [];
```

Comment:
I think you should use `URLSearchParams` which accept `string` and is a lot safer than regex

```js
const params = new URLSearchParams("a=1&b=2");
// URLSearchParams { a → "1", b → "2" }
[...params.values()] //[ ["a", "1" ], ["b", "2" ] ]
```